lua/null-ls/generators.lua

```diff
@@ -57,6 +57,14 @@ M.run = function(generators, params, opts, callback)
             local ok, results = protected_call(to_run, copied_params)
             a.util.scheduler()
 
+            -- filter diagnostics results with the diagnostic_filter option
+            local filter = opts.filter
+            if filter and results then
+                results = vim.tbl_filter(function(result)
+                    return filter(result, generator)
+                end, results)
```

Comment:
```suggestion
                results = vim.tbl_filter(filter, results)
```

With the other suggested changes, we can make this into a more straightforward predicate (unless you think there's a reason to access `generator` when filtering, but I'd rather avoid a chance to accidentally mutate it).